lib/Mailer/SendGrid.php

```diff
@@ -25,128 +26,114 @@ class SendGrid extends Mailer
+        if (empty($personalization['bcc'])) {
+            unset($personalization['bcc']);
         }
 
-        $smtpApi = [];
+        $content = [
+            ['type' => 'text/plain', 'value' => $message->getText()],
+            ['type' => 'text/html', 'value' => $message->getHtml()],
+        ];
 
-        if ($message instanceof TaggableInterface) {
-            $smtpApi['category'] = (array) $message->getTag();
+        foreach ($content as $i => $c) {
+            if (empty($c['value'])) {
+                unset($content[$i]);
+            }
         }
 
-        if ($message instanceof MetadataAwareInterface) {
-            $metadata = array_filter($message->getMetadata());
-            // Do not set an empty array on unique_args. json_encode turns it into []; SendGrid expects {}.
-            if (count($metadata)) {
-                $smtpApi['unique_args'] = $metadata;
-            }
+        $parameters = [
+            'personalizations' => [$personalization],
+            'from' => $this->formatRecipients($message->getFrom())[0],
+            'reply_to' => $this->formatRecipients($message->getReplyTo()),
+            'content' => array_values($content),
+        ];
+
+        if (empty($parameters['reply_to'])) {
+            unset($parameters['reply_to']);
+        } else {
+            $parameters['reply_to'] = $parameters['reply_to'][0];
         }
 
-        $inline = [];
         if ($message instanceof AttachmentsAwareInterface) {
-            // Store inline attachment references
-            list(, $inline) = $this->processAttachments($message->getAttachments());
+            $attachments = $message->getAttachments();
+            if (!empty($attachments)) {
+                $parameters['attachments'] = $this->processAttachments($attachments);
+            }
         }
 
-        $parameters = [
-            'api_user' => $username,
-            'api_key'  => $password,
-            'to'       => $toEmails,
-            'toname'   => $toNames,
-            'from'     => $from->getEmail(),
-            'fromname' => $from->getName(),
-            'subject'  => $message->getSubject(),
-            'text'     => $message->getText(),
-            'html'     => $message->getHtml(),
-            'bcc'      => $bccEmails,
-            'replyto'  => $message->getReplyTo(),
-            'content'  => $inline,
-        ];
-
-        if ($headers = $message->getHeaders()) {
-            $parameters['headers'] = json_encode($headers);
+        if ($message instanceof TaggableInterface) {
+            $parameters['categories'] = (array) $message->getTag();
         }
 
-        if ($smtpApi) {
-            $parameters['x-smtpapi'] = json_encode(array_filter($smtpApi));
+        if ($message instanceof MetadataAwareInterface) {
+            $metadata = array_filter($message->getMetadata());
+            if (!empty($metadata)) {
+                $parameters = array_merge($parameters, $metadata);
```

Comment:
This looks wrong to me. AFAICT, custom metadata should go in the `custom_args` parameter, not on top-level